src/proto/generation.proto

```diff
@@ -240,6 +241,92 @@ message ClassifierParameters {
   optional Action realized_action = 3;
 }
 
+//
+// Transforms
+//
+
+enum BorderMode {
+  BORDER_REFLECT = 0;   // reflect image values across the border
+  BORDER_REPLICATE = 1; // replicate border values outside the image
+  BORDER_WRAP = 2;      // wrap around / tile the image values
+  BORDER_ZERO = 3;      // use 0 for locations outside the image
+}
+
+enum ColorMatchMode {
+  COLOR_MATCH_HSV = 0;  // match hue, saturation, and value histograms
+  COLOR_MATCH_LAB = 1;  // match lightness, a, and b histograms
+  COLOR_MATCH_RGB = 2;  // match red, green, and blue histograms
+}
+
+message TransformAddNoise {
+  float amount = 1; // amount of noise to add
+  uint32 seed = 2;  // random seed for noise generation
+}
+
+message TransformBlend {
+  float amount = 1;     // blend amount from 0 to 1
+  Artifact target = 2;  // target image
+}
+
+message TransformColorMatch {
+  ColorMatchMode color_mode = 1;
+  Artifact image = 2;
+}
+
+message TransformContrast {
+  optional float brightness = 1;  // scale on pixel intensities, 1.0 is no change
+  optional float contrast = 2;    // contrast adjustment, 1.0 is no change
+}
+
+message TransformDepthCalc {
+  optional bool export = 1;        // return the depth map as an artifact
+  optional float blend_weight = 2; // blend factor between AdaBins (0.0) and MiDaS (1.0)
+}
+
+message TransformWarp2d {
+  BorderMode border_mode = 1; // extrapolation of border pixels
+  float translate_x = 4;      // horizontal translation in pixels
+  float translate_y = 5;      // vertical translation in pixels
+  float rotate = 2;           // rotation in degrees
+  float scale = 3;            // scale factor
+}
+
+message TransformWarp3d {
+  BorderMode border_mode = 1; // extrapolation of border pixels
+  float translate_x = 3;      // horizontal translation
+  float translate_y = 4;      // vertical translation
+  float translate_z = 5;      // forward translation
+  float rotate_x = 6;         // rotation around X-axis in degrees (pitch)
+  float rotate_y = 7;         // rotation around Y-axis in degrees (yaw)
+  float rotate_z = 8;         // rotation around Z-axis in degrees (roll)
+  float near_plane = 9;       // near clipping plane of view frustum
+  float far_plane = 10;       // far clipping plane of view frustum
+  float fov = 11;             // field of view in degrees
+}
+
+message TransformWarpFlow {
+  optional Artifact flow_map = 1;   // precomputed flow asset
+  optional Artifact prev_frame = 2;
+  optional Artifact next_frame = 3;
+}
+
+message TransformOperation {
+  oneof xform {
+    TransformAddNoise add_noise = 1;
+    TransformBlend blend = 6;
+    TransformColorMatch color_match = 2;
+    TransformContrast contrast = 8;
+    TransformDepthCalc depth_calc = 7;
+    TransformWarp2d warp2d = 3;
+    TransformWarp3d warp3d = 4;
+    TransformWarpFlow warp_flow = 5;
+  }
+}
+
+message TransformSequence {
+  repeated TransformOperation operations = 1;
```

Comment:
This appears to be an attempt to reproduce the functionality of Chains without chains?

Reply:
With chains the image data would need to be repeatedly encoded/decoded from PNG. As transform sequence they can be held in RAM to do the flexible series of transformations plus the one heavier model inference (warp2d, warp3d, or warp_flow).

Reply:
@pharmapsychotic Here's the problem with your assumption.

That it needs to consume or produce PNGs.

Reply:
I do agree that this is re-inventing chains. If these are all lighter operations, ie. not a bunch of ML models, it could make sense to do them in sequence inside the same call for the middleware. However, I think it would make sense for that to be a future optimization.

The middleware could auth-batch a bunch of operations when it sees them going to the same backend in sequence. It could send a `Chain` call to the backend. This way you are not inventing "yet another chain", and it keeps flexibility to de-couple operations from `TransformService` that really should be in separate backends, such as ML models. By doing a `TransformService` only chain-implementation, you are making composability immensely harder and will eventually end up with a huge `TransformService` that has a terrible usage profile instead of decoupling it.